engines/nancy/graphics.cpp

```diff
@@ -0,0 +1,263 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+#include "engines/nancy/graphics.h"
+
+#include "engines/nancy/renderobject.h"
+#include "engines/nancy/resource.h"
+#include "engines/nancy/nancy.h"
+#include "engines/nancy/resource.h"
+#include "engines/nancy/cursor.h"
+#include "engines/nancy/state/scene.h"
+#include "engines/nancy/ui/viewport.h"
+
+#include "common/file.h"
+#include "common/system.h"
+
+#include "image/bmp.h"
+
+#include "engines/util.h"
+
+namespace Nancy {
+
+const Graphics::PixelFormat GraphicsManager::inputPixelFormat = Graphics::PixelFormat(2, 5, 5, 5, 0, 10, 5, 0, 0);
+const Graphics::PixelFormat GraphicsManager::screenPixelFormat = Graphics::PixelFormat(2, 5, 6, 5, 0, 11, 5, 0, 0);
+const Graphics::PixelFormat GraphicsManager::clut8Format = Graphics::PixelFormat::createFormatCLUT8();
+
+void GraphicsManager::init() {
+    initGraphics(640, 480, &screenPixelFormat);
+    _screen.create(640, 480, screenPixelFormat);
+    _screen.setTransparentColor(getTransColor());
+    _screen.clear();
+
+    Common::SeekableReadStream *ob = g_nancy->getBootChunkStream("OB0");
+    ob->seek(0);
+
+    g_nancy->resource->loadImage(ob->readString(), object0);
+
+    loadFonts();
+}
+
+void GraphicsManager::draw() {
+    for (auto it : _objects) {
+        RenderObject &current = *it;
+
+        current.updateGraphics();
+
+        if (current._isVisible && current._needsRedraw) {
+            // object is visible and updated
+
+            if (current._redrawFrom) {
+                if (current.hasMoved() && !current.getPreviousScreenPosition().isEmpty()) {
+                    // Redraw previous location if moved
+                    blitToScreen(*current._redrawFrom, current.getPreviousScreenPosition());
+                }
+
+                if (current._drawSurface.hasTransparentColor()) {
+                    // Redraw below if transparent
+                    blitToScreen(*current._redrawFrom, current.getScreenPosition());
+                }
+            }
+
+            // Draw the object itself
+            blitToScreen(current, current.getScreenPosition());
+        } else if (!current._isVisible && current._needsRedraw && current._redrawFrom && !current.getPreviousScreenPosition().isEmpty()) {
+            // Object just turned invisible, redraw below
+            blitToScreen(*current._redrawFrom, current.getPreviousScreenPosition());
+        }
+
+        current._needsRedraw = false;
+        current._previousScreenPosition = current._screenPosition;
+    }
+
+    // Draw the screen
+    _screen.update();
+}
+
+void GraphicsManager::addObject(RenderObject *object) {
+    for (auto &r : _objects) {
```

Comment:
Could use a `const auto &` on loops like this, which theoretically helps the compiler, although I don't know how much difference it makes in practice.

Reply:
Thanks, I've added const to this and a few loops like it.